crates/plugins/mediator-coordination/src/forward/routing.rs

```diff
@@ -0,0 +1,132 @@
+use didcomm::{Message, PackEncryptedOptions, UnpackOptions};
+
+use crate::{
+    model::stateful::entity::Messages,
+    web::{error::MediationError, AppState, AppStateRepository},
+};
+/// mediator receives messages of type forward then it unpacks the messages and stores it for pickup
+/// the unpacked message is then repacked for further transmission.
+/// Note: Stored messages are not re_packed and must be before transmission in case of
+/// Rewrapping.
+pub async fn mediator_forward_process(
+    payload: &str,
+    state: &AppState,
+) -> Result<Vec<String>, MediationError> {
+    // unpack encrypted payload message
+    let AppStateRepository {
+        message_repository, ..
+    } = state
+        .repository
+        .as_ref()
+        .expect("Missing Persistence Layer");
+
+    let mut forward = Vec::new();
+    let mediator_did = &state.diddoc.id;
+    let (unpack_msg, _) = Message::unpack(
+        payload,
+        &state.did_resolver,
+        &state.secrets_resolver,
+        &UnpackOptions::default(),
+    )
+    .await
+    .map_err(|_| MediationError::MessageUnpackingFailure)?;
+    if unpack_msg.to.is_some() {
+        let dids = unpack_msg.clone().to.expect("to field is None");
+        for did in dids {
+            let (re_packed_msg, _) = unpack_msg
+                .pack_encrypted(
+                    &did,
+                    Some(mediator_did),
+                    None,
+                    &state.did_resolver,
+                    &state.secrets_resolver,
+                    &PackEncryptedOptions::default(),
+                )
+                .await
+                .expect("could not pack message: {0}");
+
+            let messages = Messages {
+                id: None,
+                message: unpack_msg.clone(),
+                recipient_did: did,
+            };
+            message_repository
+                .store(messages)
+                .await
+                .map_err(|_| MediationError::PersisenceError)
+                .unwrap();
+            forward.push(re_packed_msg)
+        }
```

Comment:
According to the [routing protocol 2.0](https://identity.foundation/didcomm-messaging/spec/v2.1/#routing-protocol-20):
> There are 3 roles in the protocol: sender, mediator, and recipient. The sender emits messages of type **forward** to the mediator. The mediator **unpacks (decrypts)** the **payload** of an encrypted forward message and passes on the result (a blob that probably contains a differently encrypted payload) to the recipient.
>> Mediator process:
>> 1. Receive ‘forward’ message.
>> 2. Retrieve service endpoint pre-configured by recipient (**next** attribute).
>> 3. Transmit payload message to service endpoint in the manner specified in the [transports] section.

The mediator will receive an encrypted payload (a forward message from the sender), decrypts it and send it to the recipient. So you need to **unpack** the message first, look at the **next** field (which is the recipient's did or key) of the **body** section of the message, and then store that did (or key) along with the **attachments** of the message.
> Normally, the payload attached to the forward message received by the mediator is transmitted directly to the receiver with no further packaging. However, optionally, the mediator can attach the opaque payload to a new forward message (appropriately anoncrypted), which then acts as a fresh outer envelope for the second half of the delivery. This rewrapping means that the “onion” of packed messages stays the same size rather than getting smaller as a result of the forward operation.
Rewrapping mode is invisible to senders, but mediators need to know about it, since they change their behavior as a result. Receivers also need to know about it, because it causes them to receive a double-packaged message instead of a singly-packaged one. The outer envelope is a forward message where to is the receiver itself.

To support this, we can modify the signature of the **`mediator_forward_process`** function to include an additional parameter that optionally enables the rewrapping process of the attached payload before storing it. To wrap the payload, you can use the `wrap_in_forward` function from the `didcomm` crate, located at `didcomm::protocols::routing::wrap_in_forward`.

Additionally, avoid directly unwrapping results whenever possible to prevent potential runtime panics and facilitate debugging.

Reply:
yes i agree with you, i thought of this too but there are somme points i want to bring out
- you remember the first session with @francis-pouatcha where i said the **body** field of the didcomm::Message struct does not has any **next** attribute if you check on that.

Reply:
The content of the `body` field depends on the `type` attribute of the DIDComm message and must be a JSON object.
> body - OPTIONAL. The body attribute contains all the data and structure that are uniquely defined for the schema associated with the `type` attribute. If present, it MUST be a JSON object that conforms to [RFC 7159](https://datatracker.ietf.org/doc/html/rfc7159).

Since the `next` attribute is a String, you need to check it within the `String(String)` variant of the JSON `Value` type from `didcomm`.

Reply:
Thanks for the hint :pray: